pkg/datastore/pubsub/mqtt/mqtt.go

```diff
@@ -0,0 +1,193 @@
+package mqtt
+
+import (
+	"encoding/json"
+	"strconv"
+	"time"
+
+	"gofr.dev/pkg"
+	"gofr.dev/pkg/datastore"
+	"gofr.dev/pkg/datastore/pubsub"
+	"gofr.dev/pkg/errors"
+	"gofr.dev/pkg/gofr/types"
+	"gofr.dev/pkg/log"
+
+	mqtt "github.com/eclipse/paho.mqtt.golang"
+)
+
+type MQTT struct {
+	Client mqtt.Client
+	logger log.Logger
+	config *Config
+}
+
+type Config struct {
+	Protocol                string
+	Hostname                string
+	Port                    int
+	Username                string
+	Password                string
+	ClientID                string
+	Topic                   string
+	QoS                     byte
+	Order                   bool
+	ConnectionRetryDuration int
+}
+
+// New establishes connection to Kafka using the config provided in KafkaConfig
+func New(config *Config, logger log.Logger) (pubsub.PublisherSubscriber, error) {
+	options := mqtt.NewClientOptions()
+	options.AddBroker("tcp://" + config.Hostname + ":" + strconv.Itoa(config.Port))
+	options.SetClientID(config.ClientID)
+
+	if config.Username != "" {
+		options.SetUsername(config.Username)
+	}
+
+	if config.Password != "" {
+		options.SetPassword(config.Password)
+	}
+
+	options.SetOrderMatters(config.Order)
+
+	// upon connection to the client, this is called
+	options.OnConnect = func(client mqtt.Client) {
+		logger.Debug("Connected")
+	}
+
+	// this is called when the connection to the client is lost; it prints "Connection lost" and the corresponding error
+	options.OnConnectionLost = func(client mqtt.Client, err error) {
+		logger.Errorf("Connection lost: %v", err)
+	}
+
+	options.ConnectRetryInterval = time.Second * time.Duration(config.ConnectionRetryDuration)
+
+	// create the client using the options above
+	client := mqtt.NewClient(options)
+
+	if token := client.Connect(); token.Wait() && token.Error() != nil {
+		logger.Errorf("cannot connect to MQTT, HostName : %v, Port : %v, error : %v", config.Topic, config.Port, token.Error())
```

Comment:
In this error log, config.Topic is used for the value of HostName, isn't it should be Topic?